datajunction-server/datajunction_server/sql/parsing/ast.py

```diff
@@ -1119,12 +1119,65 @@ def add_table(self, table: "Table") -> "Wildcard":
         return self
 
     def __str__(self) -> str:
-        return "*"
+        return (self.namespace[0].name + SEPARATOR if self.namespace else "") + "*"
 
     @property
     def type(self) -> ColumnType:
         return WildcardType()
 
+    async def compile(self, ctx: CompileContext):
+        """
+        Compile a Wildcard AST node. If the wildcard is used in a SELECT statement, we
+        replace it with the equivalent of explicitly selecting all upstream columns.
+        """
+        if isinstance(self.parent, Select):
+            wildcard_parent = cast(Select, self.parent)
+
+            # If a wildcard is used in a SELECT statement, we pull the columns that it
+            # represents by scanning all relations in the FROM clause (both the primary table
+            # and any joined tables)
+            wildcard_table_namespace = (
+                self.namespace[0].name
+                if self.namespace and isinstance(self.namespace, list)
+                else None
+            )
+            from_relations = [
+                wildcard_parent.from_.relations[0].primary,
+                *[ext.right for ext in wildcard_parent.from_.relations[0].extensions],
+            ]
+
+            for relation in from_relations:
+                await relation.compile(ctx)
+                if (
+                    not wildcard_table_namespace
+                    or relation.alias_or_name.name == wildcard_table_namespace
+                ):
+                    # Figure out where the relation's columns are stored depending on the relation type
+                    if isinstance(relation, Table):
+                        wildcard_origin = cast(Table, relation)
+                        if wildcard_origin_node := wildcard_origin.dj_node:
+                            relation_columns = wildcard_origin_node.columns
+                        else:
+                            relation_columns = wildcard_origin._cte_columns
+                    else:
+                        wildcard_origin = cast(Query, relation)
+                        relation_columns = wildcard_origin.select.projection
+
+                    # Use these columns to replace the wildcard
+                    for col in relation_columns:
+                        wildcard_parent.projection.append(
+                            Column(
+                                name=Name(col.name)
+                                if isinstance(col.name, str)
+                                else col.name,
+                                _table=wildcard_origin,
+                                _type=col.type,
+                            ),
+                        )
+            wildcard_parent.projection.remove(self)
+        else:
```

Comment:
nit: it may be easier to invert this conditional:
```
if not isinstance(self.parent, Select):
    return super().compile(ctx)

# and then the main code w/o the need to be indented
```

Reply:
Good point!